### ISSUE TITLE
Stops after "Relation # is not a valid multipolygon (built 0 complete rings, but * segments are unmatched)"

### ISSUE THREAD

**Issue by JaffaKetchup** [dfyz/osm-renderer#13]
Hey @dfyz,

I've been looking for a project like this for a while, so I'm glad to have found one!

I am having some difficulties however.
I've tried both the Great Britain and Croatia Geofabrik extracts, and they both eventually output a line as is in the title.
I'm sure this is normal, but the program never seems to recover or crash - it just pauses forever. I left the GB extract for over 2 hours at this stage and there were no more events.
I have 32GB RAM, which may not be enough for GB, but it should definitely be enough for Croatia.

Not sure if you still are maintaining this project, but I do hope so!

**Comment by dfyz** [comment]
**tl;dr** This was a performance bug, which should now hopefully be fixed.

Hey @JaffaKetchup,

I'm glad you find this project useful! It is somewhat maintained, meaning that I fix bugs and update dependencies whenever I have some free time on my hands.

What you stumbled upon with Croatia was a latent performance issue in [multipolygon ring detection](https://wiki.openstreetmap.org/wiki/Relation:multipolygon#Examples_in_XML). The algorithm was quadratic in the number of nodes for invalid multipolygons. Unfortunately, for Croatia, some huge multipolygons in the Geofabrik extracts are, in fact, invalid because some of their rings partially belong to other countries and are not exported. The worst offenders were [Istra](https://www.openstreetmap.org/relation/8063578) (tens of thousands of nodes, some of which belong to Slovenia and Italy) and [the Adriatic Sea](https://www.openstreetmap.org/relation/9350128) (hundreds of thousands of nodes, which obviously don't all belong to Croatia).

I re-wrote the algorithm to make it linear, and now I can successfully import and render Croatia on my laptop. I presume that Great Britain had the same problem, but unfortunately I don't think my laptop has enough memory to test the fix for the UK.

Also, just to make expectations straight: the renderer is intended to be used with small (e.g., city-sized) regions. Nothing stops you from importing larger regions (e.g., countries) if you want to, and if it works, it's great! However, some of the design decisions I took make this render a suboptimal choice in this case. For one thing, the importer is extremely memory-hungry so you might run out of memory even with plenty of RAM; for another, the imported map might end up larger even than the raw OSM file.

**Comment by JaffaKetchup** [comment]
Hey @dfyz,

That's great news, thanks for looking into this!
I'll see if I can re-test at some point this week.

Is there any way I can donate to you (no bitcoin)?

**Comment by JaffaKetchup** [comment]
This is now fixed, thanks a lot!

Croatia now processes, but I don't have enough RAM to complete GB (although it does get further along than before).
